README.md

# generations

Convert birth years to generation names.

This package contains functions that will convert years of birth into
names of generational cohorts. For example, `1989` becomes
`"Millennial"`. The package contains several helper functions to modify
the underlying lookup table as well as way to query it.

## Installation

The package currently lives on GitHub, so you can install it like you
would with any other package on GitHub:

``` r
remotes::install_github("joeystanley/generations")
```

You can then load it like you can with any library.

``` r
library(generations)
```

For the purposes of this tutorial, I’ll load `ggplot` as well.

``` r
library(dplyr)
library(ggplot2)
```

## Converting years to generations

The main function in this package is `generations()`. Given a vector of
integers, it’ll return a factor of generation names. First, I’ll
generate some random years of birth.

``` r
yobs <- floor(runif(10, 1900, 2020))
yobs
```

    ##  [1] 1908 1998 2013 1932 1920 1904 1921 1976 1902 1900

I can now easy convert that into generations.

``` r
generations(yobs)
```

    ##  [1] G.I.       Millennial Gen Z      Silent     G.I.       Lost      
    ##  [7] G.I.       Gen X      Lost       Lost      
    ## Levels: Lost G.I. Silent Gen X Millennial Gen Z

This function works on any year betwen 1435 and 2030. Numbers outside
that range return `NA`.

Note that by default, the function will return the vector as *factor*,
with the levels ordered so that the oldest generation in the vector is
first. To get a character vector instead, add the argument `as_factor =
FALSE`.

### Customizing output

There are some tweaks you can do to adjust the output of
`generations()`. First, you can return longer forms of the generational
names by specifying `full_names = TRUE`.

``` r
generations(yobs, full_names = TRUE)
```

    ##  [1] G.I. Generation       Millennial Generation Generation Z         
    ##  [4] Silent Generation     G.I. Generation       Lost Generation      
    ##  [7] G.I. Generation       Generation X          Lost Generation      
    ## [10] Lost Generation      
    ## 6 Levels: Lost Generation G.I. Generation Silent Generation ... Generation Z

What this does is simply add `"Generation"` to the end of each one,
unless it’s `"Gen X"` (or Y, or Z), in which case it’ll expand it out to
simply `"Generation X"`.

You can also show the years included in each generation by adding the
`years = TRUE` argument. This will add a space and, inside a pair of
parentheses, the start and end years of that generation, separated by an
en dash.

``` r
generations(yobs, years = TRUE)
```

    ##  [1] G.I. (1908–1928)       Millennial (1984–2007) Gen Z (2008–2030)     
    ##  [4] Silent (1929–1945)     G.I. (1908–1928)       Lost (1886–1907)      
    ##  [7] G.I. (1908–1928)       Gen X (1964–1983)      Lost (1886–1907)      
    ## [10] Lost (1886–1907)      
    ## 6 Levels: Lost (1886–1907) G.I. (1908–1928) ... Gen Z (2008–2030)

The primary purpose of this is for visualizations, since not everyone is
familiar with (or agrees with) the year ranges. For example, if you’ve
got a bunch of people and want to visualize the distribution of when
they were born, you could have very informative legends.

``` r
many_yobs <- tibble(yob = floor(rnorm(1000, 1975, 15))) %>%
  mutate(gen = generations(yob, full_names = TRUE, years = TRUE))
ggplot(many_yobs, aes(yob, fill = gen)) + 
  geom_histogram(binwidth = 1) + 
  scale_fill_brewer(name = NULL, palette = "Set1")
```

![](man/figures/README-unnamed-chunk-10-1.png)<!-- -->

How this additional portion is formatted can be adjusted. If rendering
an en dash is troublesome for you, you can change it to something else
with `years_range_sep`. You may also want to change the space between
the generation name and the opening parenthesis into a newline character
with `years_sep`, again for visualization purposes.

``` r
many_yobs <- many_yobs %>%
  mutate(gen = generations(yob, full_names = TRUE, years = TRUE,
                           years_sep = "\n", years_range_sep = " to "))

ggplot(many_yobs, aes(yob, fill = gen)) + 
  geom_histogram(binwidth = 1) + 
  scale_fill_brewer(name = NULL, palette = "Set1") + 
  labs(x = NULL) + 
  theme(legend.key.height = unit(1, "cm"))
```

![](man/figures/README-unnamed-chunk-11-1.png)<!-- -->

If you want to get really fancy, you can make the legend keys
approximate the width they take up on the *x*-axis and put better tics
marks.

``` r
widths <- many_yobs %>%
  group_by(gen) %>%
  summarize(width = max(yob) - min(yob)) %>%
  ungroup() %>%
  mutate(width = width / max(width) * 1.4) # you may have to fudge this a little more

ggplot(many_yobs, aes(yob, fill = gen)) + 
  geom_histogram(binwidth = 1) + 
  scale_fill_brewer(name = NULL, palette = "Set1") + 
  scale_x_continuous(breaks = c(1929, 1946, 1964, 1984, 2008, 2030)) + 
  labs(x = NULL) + 
  theme(legend.position = "bottom") + 
  guides(fill = guide_legend(nrow = 1, label.position = "bottom", 
                             keywidth = widths$width, default.unit = "inches"))
```

![](man/figures/README-unnamed-chunk-12-1.png)<!-- -->

## Querying generation data

To see a list of the generational data, you can use
`show_generations()`, which will return a data frame containing the
names, start years, and end years.

``` r
show_generations()
```

    ##              name start  end
    ## 1           Gen Z  2008 2030
    ## 2      Millennial  1984 2007
    ## 3           Gen X  1964 1983
    ## 4          Boomer  1946 1963
    ## 5          Silent  1929 1945
    ## 6            G.I.  1908 1928
    ## 7            Lost  1886 1907
    ## 8      Missionary  1865 1885
    ## 9     Progressive  1844 1864
    ## 10         Gilded  1822 1843
    ## 11 Transcendental  1794 1821
    ## 12     Compromise  1773 1793
    ## 13     Republican  1746 1772
    ## 14        Liberty  1727 1745
    ## 15      Awakening  1704 1726
    ## 16  Enlightenment  1675 1703
    ## 17       Glorious  1649 1674
    ## 18       Cavalier  1621 1648
    ## 19        Puritan  1594 1620
    ## 20  Parliamentary  1569 1593
    ## 21    Elizabethan  1542 1568
    ## 22       Reprisal  1517 1541
    ## 23    Reformation  1497 1516
    ## 24       Humanist  1459 1496
    ## 25     Aurthurian  1435 1458

You can also get simple information. For example, if you want to know
when the start or end year of a particular generation is, you can use
`get_start()` or `get_end()`:

``` r
get_start("Silent")
```

    ## [1] 1929

``` r
get_end("Millennial")
```

    ## [1] 2007

You can also find the names of neighboring generations with
`get_prev_gen()` and `get_next_gen()`, though these were mostly created
for internal purposes only rather than for you to use.

``` r
get_next_gen("Millennial")
```

    ## [1] "Gen Z"

``` r
get_prev_gen("Missionary")
```

    ## [1] "Progressive"

Note that if ask for something newer than Gen Z or older than Aurthurian
it will return `NA`.

## Customizing generation data

The data that this package uses is loaded as a hidden object when you
load the package. You may modify it with the functions described in this
section. These changes will affect the dataset so long as the
`generations` package is loaded. You’ll have to reset the data each time
to reload it.

The labels and years for each generation are mostly borrowed from Howe &
Strauss’ Generational Theory books. However, not everyone agrees on the
names and year ranges for the various generations. For this reason, the
`generations` package makes it easy to modify the generations data to
your liking.

To rename a generation, use `rename_generation()`, with the old name
first and the new name second. For example, if you want to use *Zoomer*
instead of *Gen Z*, you can do so.

``` r
rename_generation("Gen Z", "Zoomer")
```

    ## Gen Z has been renamed Zoomer

You’ll get a message informing you that the change has been made. If you
now run `show_generations()` you’ll see that the change has been made
and if you rerun `generations()`, you’ll get updated results.

``` r
show_generations()
```

    ##              name start  end
    ## 1          Zoomer  2008 2030
    ## 2      Millennial  1984 2007
    ## 3           Gen X  1964 1983
    ## 4          Boomer  1946 1963
    ## 5          Silent  1929 1945
    ## 6            G.I.  1908 1928
    ## 7            Lost  1886 1907
    ## 8      Missionary  1865 1885
    ## 9     Progressive  1844 1864
    ## 10         Gilded  1822 1843
    ## 11 Transcendental  1794 1821
    ## 12     Compromise  1773 1793
    ## 13     Republican  1746 1772
    ## 14        Liberty  1727 1745
    ## 15      Awakening  1704 1726
    ## 16  Enlightenment  1675 1703
    ## 17       Glorious  1649 1674
    ## 18       Cavalier  1621 1648
    ## 19        Puritan  1594 1620
    ## 20  Parliamentary  1569 1593
    ## 21    Elizabethan  1542 1568
    ## 22       Reprisal  1517 1541
    ## 23    Reformation  1497 1516
    ## 24       Humanist  1459 1496
    ## 25     Aurthurian  1435 1458

``` r
generations(yobs)
```

    ##  [1] G.I.       Millennial Zoomer     Silent     G.I.       Lost      
    ##  [7] G.I.       Gen X      Lost       Lost      
    ## Levels: Lost G.I. Silent Gen X Millennial Zoomer

Because many people may want to use the term *Zoomer* instead of *Gen
Z*, a shortcut function, `use_zoomer()`, which is just a wrapper around
`rename_generation("Gen Z", "Zoomer")`, is included in the package. The
other shortcut functions are `use_gen_y()`, `use_13th()`,
`use_baby_boom()` as well as their reciprocals `use_gen_z()`,
`use_millennial()`, `use_gen_x()` and `use_boomer()`.

You may also want to change the years. For example, many people consider
1997 as the end of the Millennial Generation. You can make this change
with `redefine_generation()`. With this function, you must specify the
new start and the new end year.

``` r
redefine_generation("Millennial", 1983, 1997)
```

    ## Gen X is now from 1964 to 1982

    ## Millennial is now from 1983 to 1997

    ## Zoomer is now from 1998 to 2030

Since changing one generation impacts adjacent generations, you’ll get a
message showing you what the new ranges are for this, the previous, and
the next generations.

You can reset the data back to its original form with
`reset_generations()`.

## Conclusion

That’s the package so far\! I plan on adding more things in the future,
primarily to handle stability issues and to include some error catching.
Hopefully, if you use generational cohorts in your data, this package is
useful for you.


NEWS.md
# generations 0.1.0

* Full package version is ready to go.

# generations 0.0.0.9000

* Initial commit.


R/getters.R
#' Show generational data
#' 
#' Displays the generation names, start years, and end years.
#' @examples 
#' show_generations()
#' @export
show_generations <- function() {
  .gen_df[order(-.gen_df$start),c("name", "start", "end")]
}

#' Get a generation's start or end year
#' 
#' @param generation a string, exactly matching a generation's name
#' 
#' @examples 
#' get_start("Millennial")
#' get_start(c("Millennial", "Lost"))
#' get_end("Millennial")
#' get_end(c("Millennial", "Lost"))
#' @export
get_start <- function(generation) {
  indices <- match(generation, .gen_df$name)
  .gen_df$start[indices]
}

#' @rdname get_start
#' @export
get_end <- function(generation) {
  indices <- match(generation, .gen_df$name)
  .gen_df$end[indices]
}





#' Get the previous generation
#'
#' Given a generation's name, find what the previous or next generation is.
#' @param gen a string, exactly matching a generation's name
#' @return a string. If the generation is out of range (such as the generation
#' after Gen Z, it returns \code{NA}.)
#' @examples
#' get_prev_gen("Millennial")
#' get_prev_gen("Aurthurian")
#' get_next_gen("Silent")
#' get_next_gen("Gen Z")
#' @export
get_prev_gen <- function(gen) {
  prev_gen_rank <- match(gen, .gen_df$name) + 1
  if (prev_gen_rank > 0 & prev_gen_rank <= length(.gen_df$name)) {
    .gen_df$name[prev_gen_rank]
  } else {
    NA
  }
}


#' @rdname get_prev_gen
#' @export
get_next_gen <- function(gen) {
  next_gen_rank <- match(gen, .gen_df$name) - 1
  if (next_gen_rank > 0 & next_gen_rank <= length(.gen_df$name)) {
    .gen_df$name[next_gen_rank]
  } else {
    NA
  }
}


R/setters.R
#' Reset underlying generational data
#' 
#' Changes to the underlying dataset can be made with \code{rename_generation}
#' or \code{redefine_generation}. This function undoes all these changes and 
#' reverts the data back to its out-of-the-box names and definitions.
#' 
#' @seealso \code{\link{rename_generation}}, \code{\link{redefine_generation}}
#' 
#' @examples 
#' reset_generations()
#' 
#' @export
reset_generations <- function() {
  .gen_df <<- assign(".gen_df", generations::gen_df, envir = globalenv())
}




#' Rename a generation
#' 
#' Not all people refer to generations the same way. The names I have chosen 
#' come from Howe & Strauss' book, though I have taken some liberty with the 
#' most recent four generations. This family of functions lets you rename
#' a generation to fit your personal tastes.
#' 
#' The main function is \code{rename_generation}. All others are shortcuts for
#' some of the more common changes I expect people to make.
#' 
#' The changes here modify the underlying data and will be reflected in any
#' subsequent reference to the generational data. You can view what the current 
#' dataset is with \code{show_generations}. Any change you make can be reset using 
#' \code{reset_generations}.
#' 
#' @param old_name a string, exactly matching a generation's name
#' @param new_name a string, representing what you want that generation to be 
#' called
#' 
#' @examples 
#' rename_generation("Gen Z", "Zoomer")
#' rename_generation("Boomer", "Baby Boom")
#' show_generations()
#' 
#' use_gen_z()
#' use_zoomer()
#' use_gen_y()
#' use_millennial()
#' use_13th()
#' use_gen_x()
#' use_boomer()
#' use_baby_boom()

#' @export
rename_generation <- function(old_name, new_name) {
  .gen_df[.gen_df$name == old_name,]$name <<- new_name
  message(old_name, " has been renamed ", new_name)
}

#' @rdname rename_generation
#' @export
use_zoomer <- function() {
  rename_generation("Gen Z", "Zoomer")
}

#' @rdname rename_generation
#' @export
use_gen_z <- function() {
  rename_generation("Zoomer", "Gen Z")
}

#' @rdname rename_generation
#' @export
use_gen_y <- function() {
  rename_generation("Millennial", "Gen Y")
}

#' @rdname rename_generation
#' @export
use_millennial <- function() {
  rename_generation("Gen Y", "Millennial")
}

#' @rdname rename_generation
#' @export
use_13th <- function() {
  rename_generation("Gen X", "13th")
}

#' @rdname rename_generation
#' @export
use_gen_x <- function() {
  rename_generation("13th", "Gen X")
}

#' @rdname rename_generation
#' @export
use_baby_boom <- function() {
  rename_generation("Boomer", "Baby Boom")
}

#' @rdname rename_generation
#' @export
use_boomer <- function() {
  rename_generation("Baby Boom", "Boomer")
}



#' Redefine when a generation starts and ends
#' 
#' Not all people agree on when a generation started and ended. This is especially 
#' true with the most recent few generations. The years I have chosen 
#' come from Howe & Strauss' book. This function lets you redefine the start
#' and end years of a generation to fit your personal tastes.
#' 
#' The changes here modify the underlying data and will be reflected in any
#' subsequent reference to the generational data. You can view what the current 
#' dataset is with \code{show_generations}. Any change you make can be reset using 
#' \code{reset_generations}.
#' 
#' @param generation a string, exactly matching a generation's name
#' @param start an integer, representing the year you want \code{generation} to 
#' start at
#' @param end an integer, representing thet year you want \code{generation} to 
#' end at.
#' 
#' @examples 
#' redefine_generation("Millennial", 1980, 1997)
#' show_generations()
#' @export
redefine_generation <- function(generation, start, end) {
  
  # Fix this generation
  .gen_df[.gen_df$name == generation,]$start <<- start
  .gen_df[.gen_df$name == generation,]$end <<- end

  # Fix previous generation
  if (!is.na(get_prev_gen(generation))) {
    .gen_df[.gen_df$name == get_prev_gen(generation),]$end <<- start - 1
  }

  # Fix following generation
  if (!is.na(get_next_gen(generation))) {
    .gen_df[.gen_df$name == get_next_gen(generation),]$start <<- end + 1
  }

  message(get_prev_gen(generation), " is now from ",
          get_start(get_prev_gen(generation)), " to ",
          get_end(get_prev_gen(generation)))
  message(generation, " is now from ", start, " to ", end)
  message(get_next_gen(generation), " is now from ",
          get_start(get_next_gen(generation)), " to ",
          get_end(get_next_gen(generation)))
}


R/generations.R
#' Convert year of birth to generations
#' 
#' This is the main function in this package. It will convert one or more years
#' to the corresponding generation. Generation names and years are taken from 
#' Howe & Strauss' \href{https://en.wikipedia.org/wiki/Strauss–Howe_generational_theory}{generational theory}.
#' 
#' Note that labels are included for years between 1435 and 2030. Anything outside
#' of that range will return \code{NA}.
#' 
#' @param yob a vector of integers. Currently, integers are the only accepted 
#' input, so characters and date types must be converted first.
#' @param full_names logical. If \code{TRUE}, the generation name will typically have
#' the word "generation" at the end (e.g., "Millennial" becomes "Millennial 
#' Generation"). "Gen X" (together with Y and Z) expands to "Generation X". 
#' \code{FALSE} by default.
#' @param years logical. Should the year range be included in the output? Useful
#' for printing, plotting, or otherwise displaying. For example, "Gen X" becomes
#' "Gen X (1964–1983)". \code{FALSE} by default.
#' @param years_sep a string. If \code{years} is \code{TRUE}, what should the 
#' separator between the generation name and opening parenthesis in the year 
#' range be? A single space by default, but it may look better in a visual if 
#' a new line character (backslash + n) is used.
#' @param years_range_sep a string. If \code{years} is \code{TRUE}, what should 
#' go between the start and end years. By default, an en-dash ("–"). For more 
#' print-friendly characters, try a hyphen.
#' @param as_factor logical. Should the vector be returned as a factor? If so, 
#' the levels will be chronological order (oldest generation first). If not, it
#' will be returned as a character vector. \code{TRUE} by default.
#' 
#' @return a vector 
#'
#' @examples
#' # Generate some sample data
#' yobs <- sort(floor(runif(10, 1880, 2020)))
#' 
#' # Convert to the generations
#' generations(yobs)
#' generations(yobs, full_names = TRUE)
#' generations(yobs, full_names = TRUE, years = TRUE)
#' generations(yobs, years = TRUE, years_range_sep = " to ")
#' 
#' @export
generations <- function(yob, full_names = FALSE, 
                        years = FALSE, years_sep = " ", years_range_sep = "–",
                        as_factor = TRUE) {
  
  # https://jennybc.github.io/purrr-tutorial/bk01_base-functions.html#vapply()_vs_map_*()
  gens <- vapply(yob,
                 function(x) .gen_df[x >= .gen_df$start & x <= .gen_df$end,"name"][[1]],
                 character(1))
  
  if (full_names) {
    gens <- full_generation_names(gens)
  }
  
  if (years) {
    gens <- add_years(gens, years_sep, years_range_sep)
  }

  if (as_factor) {
    fct_lvls <- unique(gens[order(yob)])
    gens <- factor(gens, levels = fct_lvls)
  }

  gens
}


#' Get full generation names
#' 
#' For internal use only. Takes a vector of generation names and converts it to 
#' their long format. Specifically, it adds "Generation" to most of them, except 
#' Gen X/Y/Z, in which case it just expands the first word.
#' @param gens a vector
#' @return a vector
#' @examples 
#' full_generation_names(c("Millennial", "Gen X"))
#' @keywords internal
full_generation_names <- function(gens) {
  # Super clunky base-R version of what dplyr::case_when could do more elegantly
  
  # First handle all of them besides Gen X/Y/Z
  gens[!grepl("Gen [XYZ]", gens)] <- paste(gens[!grepl("Gen [XYZ]", gens)], "Generation")
  gens[grepl("Gen [XYZ]", gens)] <- gsub("Gen ", "Generation ", gens[grepl("Gen [XYZ]", gens)])
  gens
}

#' Get shortened generation names
#' 
#' Takes a vector of generation names and gets the shortened versions of them. 
#' Specifically, it shortens "Generation X/Y/Z" to just "Gen X/Y/Z" and in all
#' other cases it just removes the word "Generation" entirely.
#' 
#' For internal use only. Mostly used within \code{add_years()} so that 
#' \code{generations()} words properly when \code{full_names} and \code{years} 
#' are both specified.
#' @param gens a vector
#' @return a vector
#' @examples 
#' short_generation_names(c("Millennial Generation", "Generation X"))
#' @keywords internal
short_generation_names <- function(gens) {
  # Super clunky base-R version of what case_when could do more elegantly
  
  # Handle the Gen X/Y/Z ones first
  gens <- gsub("Generation ", "Gen ", gens)
  # Then just remove all " Generation"
  gens <- gsub(" Generation", "", gens)
  gens
}

#' Add years to generation name
#' 
#' Takes a vector of generation names (short or long) and two integers. Adds ` (x–y)`
#' to the end of the names. The space and en-dash can be modified with the two
#' arguments. 
#' 
#' @param gens a vector
#' @param years_sep a string
#' @param years_range_sep a string
#' @examples 
#' add_years(c("Millennial", "Gen X"))
#' @keywords internal
add_years <- function(gens, years_sep, years_range_sep) {
  paste0(gens,
         years_sep,
         "(",
         get_start(short_generation_names(gens)),
         years_range_sep,
         get_end(short_generation_names(gens)),
         ")")
}

R/display_info.R

#' Show generational data
#' 
#' Displays the generation names, start years, and end years.
#' @export
show_generations <- function() {
  gen_df %>%
    dplyr::select(name, start, end) %>%
    dplyr::arrange(-start) %>%
    print()
}


#' Get a generation's start or end year
#' 
#' @examples 
#' get_start("Millennial")
#' get_start(c("Millennial", "Lost"))
#' get_end("Millennial")
#' get_end(c("Millennial", "Lost"))
#' @export
get_start <- function(generation) {
  indices <- match(generation, gen_df$name)
  gen_df$start[indices]
}

#' @rdname get_start
#' @export
get_end <- function(generation) {
  indices <- match(generation, gen_df$name)
  gen_df$end[indices]
}





#' Get the previous generation
#'
#' Given a generation's name, find what the previous or next generation is.
#' @param gen a string, exactly matching a generation's name
#' @return a string. If the generation is out of range (such as the generation
#' after Gen Z, it returns \code{NA}.)
#' @examples
#' get_prev_gen("Millennial")
#' get_prev_gen("Aurthurian")
#' get_next_gen("Silent")
#' get_next_gen("Gen Z")
#' @export
get_prev_gen <- function(gen) {
  prev_gen_rank <- match(gen, gen_df$name) + 1
  if (prev_gen_rank > 0 & prev_gen_rank <= length(gen_df$name)) {
    gen_df$name[prev_gen_rank]
  } else {
    NA
  }
}


#' @rdname get_prev_gen
#' @export
get_next_gen <- function(gen) {
  next_gen_rank <- match(gen, gen_df$name) - 1
  if (next_gen_rank > 0 & next_gen_rank <= length(gen_df$name)) {
    gen_df$name[next_gen_rank]
  } else {
    NA
  }
}


data/gen_df_documentation.R
# Generate data
if (FALSE) {
  
  # These names and years are taken from Wikipedia: https://en.wikipedia.org/wiki/Strauss–Howe_generational_theory 
  gen_df <- data.frame(name  = rev(c("Aurthurian", "Humanist",
                                "Reformation", "Reprisal", "Elizabethan", "Parliamentary",
                                "Puritan", "Cavalier", "Glorious", "Enlightenment",
                                "Awakening", "Liberty", "Republican", "Compromise",
                                "Transcendental", "Gilded", "Progressive",
                                "Missionary", "Lost", "G.I.", "Silent",
                                "Boomer", "Gen X", "Millennial", "Gen Z")),
                      start = rev(c(1435,1459,
                                1497,1517,1542,1569,
                                1594,1621,1649,1675,
                                1704,1727,1746,1773,
                                1794,1822,1844,
                                1865,1886,1908,1929,
                                1946,1964,1984,2008)))
  gen_df$end <- c(2030, gen_df$start[-length(gen_df$start)]-1)
  gen_df <- gen_df[order(-gen_df$start),]
  print(gen_df)
  
  usethis::use_data(gen_df, overwrite = TRUE)
}